cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/EventPublishManager.java

```diff
@@ -20,37 +20,49 @@
 import com.google.inject.Inject;
 import io.cdap.cdap.common.conf.CConfiguration;
 import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.spi.events.EventWriter;
 
+import java.util.Map;
 import java.util.Set;
 
 /**
  * EventPublishManager is responsible for starting all the event publish manager threads
  */
 public class EventPublishManager extends AbstractIdleService {
 
-  private boolean publishEnabled;
-  private Set<EventPublisher> eventPublishers;
+    private final boolean publishEnabled;
+    private final Set<EventPublisher> eventPublishers;
+    private final EventWriterProvider eventWriterProvider;
 
-  @Inject
-  EventPublishManager(CConfiguration cConf, Set<EventPublisher> eventPublishers) {
-    this.publishEnabled = cConf.getBoolean(Constants.Event.PUBLISH_ENABLED);
-    this.eventPublishers = eventPublishers;
-  }
-
-  @Override
-  protected void startUp() throws Exception {
-    if (!publishEnabled) {
-      return;
+    @Inject
+    EventPublishManager(CConfiguration cConf, Set<EventPublisher> eventPublishers,
+                        EventWriterProvider eventWriterProvider) {
+        this.publishEnabled = cConf.getBoolean(Constants.Event.PUBLISH_ENABLED);
+        this.eventPublishers = eventPublishers;
+        this.eventWriterProvider = eventWriterProvider;
     }
-    eventPublishers.forEach(eventPublisher -> {
-      String eventWriterPath = eventPublisher.getEventWriterPath();
-      //TODO - load classes from the path
-
-    });
-  }
 
-  @Override
-  protected void shutDown() throws Exception {
+    @Override
+    protected void startUp() throws Exception {
+        if (!publishEnabled) {
+            return; // If publish is not enabled not to start
+        }
+        eventPublishers.forEach(eventPublisher -> {
+            // Loading the event writers from provider
+            Map<String, EventWriter> eventWriterMap = this.eventWriterProvider.loadEventWriters();
+            // Initialize the event publisher with all the event writers provided by provider
+            eventPublisher.initialize(eventWriterMap.values());
```

Comment:
This is an unchecked call. To keep the writer simple for now, we can modify the EventWriter interface to not have a generic type. Also modify the write() method parameter in EventWriter to `Collection<? extends Event<?>> events` .